openmetadata-service/src/main/java/org/openmetadata/service/security/auth/BasicAuthenticator.java

```diff
@@ -0,0 +1,477 @@
+
+  private final LoginConfiguration loginConfiguration;
+  private final boolean isEmailServiceEnabled;
+  private final boolean isSelfSignUpAvailable;
+
+  public BasicAuthenticator(
+      UserRepository userRepository, TokenRepository tokenRepository, OpenMetadataApplicationConfig config) {
+    this.userRepository = userRepository;
+    this.tokenRepository = tokenRepository;
+    this.authorizerConfiguration = config.getAuthorizerConfiguration();
+    this.loginAttemptCache = new LoginAttemptCache(config);
+    SmtpSettings smtpSettings = config.getSmtpSettings();
+    this.isEmailServiceEnabled = smtpSettings != null && smtpSettings.getEnableSmtpServer();
+    this.isSelfSignUpAvailable = config.getAuthenticationConfiguration().getEnableSelfSignup();
+    this.loginConfiguration = config.getLoginSettings();
+  }
+
+  @Override
+  public User registerUser(RegistrationRequest newRegistrationRequest) throws IOException {
+    if (isSelfSignUpAvailable) {
+      String newRegistrationRequestEmail = newRegistrationRequest.getEmail();
+      String[] tokens = newRegistrationRequest.getEmail().split("@");
+      String emailDomain = tokens[1];
+      Set<String> allowedDomains = authorizerConfiguration.getAllowedEmailRegistrationDomains();
+      if (!allowedDomains.contains("all") && !allowedDomains.contains(emailDomain)) {
+        LOG.error("Email with this Domain not allowed: " + newRegistrationRequestEmail);
+        throw new BadRequestException("Email with the given domain is not allowed. Contact Administrator");
+      }
+      validateEmailAlreadyExists(newRegistrationRequestEmail);
+      PasswordUtil.validatePassword(newRegistrationRequest.getPassword());
+      LOG.info("Trying to register new user [" + newRegistrationRequestEmail + "]");
+      User newUser = getUserFromRegistrationRequest(newRegistrationRequest);
+      // remove auth mechanism from the user
+      User registeredUser = userRepository.create(null, newUser);
+      registeredUser.setAuthenticationMechanism(null);
+      return registeredUser;
+    } else {
+      throw new CustomExceptionMessage(BAD_REQUEST, SELF_SIGNUP_ERROR);
+    }
+  }
+
+  @Override
+  public void confirmEmailRegistration(UriInfo uriInfo, String emailToken) throws IOException {
+    EmailVerificationToken emailVerificationToken = (EmailVerificationToken) tokenRepository.findByToken(emailToken);
+    if (emailVerificationToken == null) {
+      throw new EntityNotFoundException("Invalid Token. Please issue a new Request");
+    }
+    User registeredUser =
+        userRepository.get(null, emailVerificationToken.getUserId(), userRepository.getFieldsWithUserAuth("*"));
+    if (Boolean.TRUE.equals(registeredUser.getIsEmailVerified())) {
+      LOG.info("User [{}] already registered.", emailToken);
+      return;
+    }
+
+    // verify Token Expiry
+    if (emailVerificationToken.getExpiryDate().compareTo(Instant.now().toEpochMilli()) < 0) {
+      throw new RuntimeException(
```

Comment:
Lets not throw RuntimeException here. Just add a TokenExpiry error to CatalogExceptionMessage and use that